### PR TITLE
osd/PeeringState: refactor PeeringState

### PR DESCRIPTION
**Pull request by tchaikov**
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

File: src/osd/PeeringState.cc [hunk 1]
```diff
@@ -1217,120 +1209,129 @@ void PeeringState::reject_reservation()
...
+      max_les = info.last_epoch_started;
     }
   }
-  for (map<pg_shard_t, pg_info_t>::const_iterator i = infos.begin();
-       i != infos.end();
-       ++i) {
-    if (max_last_epoch_started_found <= i->second.last_epoch_started) {
-      if (min_last_update_acceptable > i->second.last_update)
-	min_last_update_acceptable = i->second.last_update;
+  for ([[maybe_unused]] auto& [shard, info] : infos) {
```
Comment (athanatos) [inline src/osd/PeeringState.cc]: If we're trying to make this more readable, perhaps s/info/candidate_info?

File: src/osd/PeeringState.cc [hunk 2]
```diff
@@ -1217,120 +1209,129 @@ void PeeringState::reject_reservation()
...
-      continue;
-    // Disqualify anyone who is incomplete (not fully backfilled)
-    if (p->second.is_incomplete())
-      continue;
-    if (best == infos.end()) {
-      best = p;
-      continue;
+  using candidate_t = typename decay_t<decltype(infos)>::value_type;
+  struct is_good {
+    // std::filterd() requires the predicate to be assignable and copy
```
Comment (athanatos) [inline src/osd/PeeringState.cc]: std::filtered

File: src/osd/PeeringState.cc [hunk 3]
```diff
@@ -1217,120 +1209,129 @@ void PeeringState::reject_reservation()
...
+    is_good(bool restrict_to_up_acting, eversion_t min_lua, epoch_t max_les,
+	    const PeeringState* thiz)
+      : restrict_to_up_acting{restrict_to_up_acting},
+	min_last_update_acceptable{min_lua}, max_les{max_les}, thiz{thiz} {}
+    is_good(const is_good& rhs) = default;
+    is_good& operator=(const is_good& rhs) = default;
+    bool operator()(const candidate_t& pi) const {
+      auto& [shard, info] = pi;
+      if (restrict_to_up_acting &&
+	  !thiz->is_up(shard) && !thiz->is_acting(shard)) {
```
Comment (athanatos) [inline src/osd/PeeringState.cc]: PeeringState * is ps elsewhere in the file iirc.

**Comment by athanatos** [comment]
I don't see any errors, but I'd feel better if this came with some unit test coverage.  calc_ec_* are static, so perhaps that's possible?

File: src/osd/PeeringState.cc (hunk 1 above)
Comment (tchaikov) [reply src/osd/PeeringState.cc]: done.

File: src/osd/PeeringState.cc (hunk 2 above)
Comment (tchaikov) [reply src/osd/PeeringState.cc]: fixed.

File: src/osd/PeeringState.cc (hunk 3 above)
Comment (tchaikov) [reply src/osd/PeeringState.cc]: updated.

**Comment by tchaikov** [comment]
i will take a closer look to see if we can have better coverage.

**Comment by tchaikov** [comment]
> I don't see any errors, but I'd feel better if this came with some unit test coverage. calc_ec_* are static, so perhaps that's possible?

@athanatos i tried to make `PeeringState::find_best_info()` a `static` method, but failed. because it has following dependencies that requires an instance of `PeeringState`

* `cct->_conf->osd_find_best_info_ignore_history_les`: this is trivial, as we can extract it out as a parameter.
* `PeeringState::pg_whoami`: we can also pass it in as a parameter
* `PeeringState::is_up()` and `PeeringState::is_acting()`: this involves more dependencies. as we need to check if a shard is in the up set or in the acting set. so we need to either pass functors doing the check into `PeeringState::find_best_info()`, or pass the references of up and acting set to it. either way, it hurts the readability.

**Comment by athanatos** [comment]
Ok, LGTM then.

**Comment by liewegas** [comment]
from sage-2019-06-20_17:09:35-rados-wip-sage-testing-2019-06-20-0926-distro-basic-smithi i see ``"2019-06-20T19:21:28.628585+0000 osd.6 (osd.6) 521 : cluster [WRN] client.4356 172.21.15.60:0/3835718446 misdirected client.4356.0:8831 pg 2.f4d6334 to osd.6 not [6,6,4] in e813/813" in cluster log``.. my guess is its this pr

**Comment by tchaikov** [comment]
i reran the one of the failed test with only the `find_best_info()` change. the test passed. see http://pulpito.ceph.com/kchai-2019-06-21_08:19:16-rados-wip-find_best_info-distro-basic-smithi/

so i think it's related to `PeeringState::init_primary_up_acting()`.

**Comment by neha-ojha** [comment]
needs rebase

**Pull request closed without merging**